File: client/src/game/systems/damage.lua
local Aspect = require 'src.ecs.aspect'
local System = require 'src.ecs.system'
local Damage = require 'src.game.components.damage'
local Fixture = require 'src.game.components.fixture'
local Health = require 'src.game.components.health'

local DamageSystem = System:new('damage', Aspect.always())

function DamageSystem:beginContact(a, b, contact)
  a = a:getUserData().entity
  b = b:getUserData().entity
  local health
  local damage

  if a:has(Damage) and b:has(Health) then
    damage = a:as(Damage)
    health = b:as(Health)
  elseif a:has(Health) and b:has(Damage) then
    health = a:as(Health)
    damage = b:as(Damage)
  end

  if health and damage then
    health.hitpoints = health.hitpoints - damage.hitpoints
  end
end

return DamageSystem

File: packages/client2/lib/game/systems/damage.dart
import 'dart:math';

import './base/collision_system.dart';
import '../../ecs/ecs.dart';
import '../box2d.dart';
import '../components/damage.dart';
import '../components/health.dart';
import '../flags.dart';

class DamageSystem extends CollisionSystem {
  static const int Flag = SystemFlags.Damage;
  int flag = DamageSystem.Flag;

  static const String Id = 'DamageSystem';
  String id = DamageSystem.Id;

  Aspect aspect = Aspect(all: [
    Health.Flag,
    Damage.Flag,
  ]);

  @override
  void update(double dt) {}

  @override
  void beginContact(ContactResult result) {
    Health healthA = result.entityA.as<Health>(Health.Id);
    Health healthB = result.entityB.as<Health>(Health.Id);

    if (healthA == null || healthB == null) {
      return;
    }

    healthA.hitpoints = max(0, healthA.hitpoints - 1);
    healthB.hitpoints = max(0, healthB.hitpoints - 1);
  }

  @override
  void endContact(ContactResult result) {}
}
